File: kt/src/test/kotlin/de/arcan/arclib/crypto/CrossValidationTest.kt
package de.arcan.arclib.crypto

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import org.junit.jupiter.api.Assertions.assertArrayEquals
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestInstance
import java.io.File

@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class CrossValidationTest {
    private val gson = Gson()
    private val hashVectors: Map<String, Any>
    private val cryptoVectors: Map<String, Any>

    init {
        val baseDir = File(System.getProperty("user.dir")).parentFile.resolve("testdata")
        hashVectors = gson.fromJson(
            baseDir.resolve("hash_vectors.json").readText(),
            object : TypeToken<Map<String, Any>>() {}.type
        )
        cryptoVectors = gson.fromJson(
            baseDir.resolve("crypto_vectors.json").readText(),
            object : TypeToken<Map<String, Any>>() {}.type
        )
    }

    private fun fromHex(hex: String): ByteArray {
        if (hex.isEmpty()) return ByteArray(0)
        return hex.chunked(2).map { it.toInt(16).toByte() }.toByteArray()
    }

    private fun toHex(bytes: ByteArray): String {
        return bytes.joinToString("") { "%02x".format(it) }
    }

    @Test
    fun `SHA-256 cross-validation`() {
        @Suppress("UNCHECKED_CAST")
        val vectors = hashVectors["sha256"] as List<Map<String, String>>
        for (v in vectors) {
            val input = fromHex(v["hex_input"]!!)
            val expected = v["output"]!!
            assertEquals(expected, toHex(Hash.sha256(input)))
        }
    }

    @Test
    fun `CRC-16-IBM cross-validation`() {
        @Suppress("UNCHECKED_CAST")
        val section = hashVectors["crc16_ibm"] as Map<String, Any>
        @Suppress("UNCHECKED_CAST")
        val vectors = section["vectors"] as List<Map<String, String>>
        for (v in vectors) {
            val input = fromHex(v["hex_input"]!!)
            val expected = Integer.parseInt(v["output"]!!.removePrefix("0x"), 16)
            assertEquals(expected, Hash.crc16(input))
        }
    }

    @Test
    fun `CRC-32-IEEE cross-validation`() {
        @Suppress("UNCHECKED_CAST")
        val section = hashVectors["crc32_ieee"] as Map<String, Any>
        @Suppress("UNCHECKED_CAST")
        val vectors = section["vectors"] as List<Map<String, String>>
        for (v in vectors) {
            val input = fromHex(v["hex_input"]!!)
            val expected = java.lang.Long.parseLong(v["output"]!!.removePrefix("0x"), 16)
            assertEquals(expected, Hash.crc32(input))
        }
    }

    @Test
    fun `HMAC-SHA256 cross-validation`() {
        @Suppress("UNCHECKED_CAST")
        val section = hashVectors["hmac_sha256"] as Map<String, Any>
        @Suppress("UNCHECKED_CAST")
        val vectors = section["vectors"] as List<Map<String, String>>
        for (v in vectors) {
            val key = fromHex(v["key"]!!)
            val data = fromHex(v["data"]!!)
            val expected = v["output"]!!
            assertEquals(expected, toHex(Hmac.sha256(key, data)))
        }
    }

    @Test
    fun `ChaCha20-Poly1305 cross-validation`() {
        @Suppress("UNCHECKED_CAST")
        val vectors = cryptoVectors["chacha20_poly1305"] as List<Map<String, String>>
        for (v in vectors) {
            val key = fromHex(v["key"]!!)
            val nonce = fromHex(v["nonce"]!!)
            val plaintext = fromHex(v["plaintext"]!!)
            val aad = fromHex(v["aad"]!!)
            val expectedCt = v["ciphertext"]!!

            val encrypted = Aead.encrypt(key, nonce, plaintext, aad)
            assertEquals(expectedCt, toHex(encrypted))

            val decrypted = Aead.decrypt(key, nonce, fromHex(expectedCt), aad)
            assertArrayEquals(plaintext, decrypted)
        }
    }

    @Test
    fun `HKDF-SHA256 cross-validation`() {
        @Suppress("UNCHECKED_CAST")
        val vectors = cryptoVectors["hkdf_sha256"] as List<Map<String, Any>>
        for (v in vectors) {
            val ikm = fromHex(v["ikm"] as String)
            val salt = (v["salt"] as String).let { if (it.isEmpty()) null else fromHex(it) }
            val info = (v["info"] as String).let { if (it.isEmpty()) null else fromHex(it) }
            val length = (v["length"] as Double).toInt()
            val expected = v["okm"] as String

            val result = Hkdf.deriveKey(ikm, salt, info, length)
            assertEquals(expected, toHex(result))
        }
    }

    @Test
    fun `Key Rotation cross-validation`() {
        @Suppress("UNCHECKED_CAST")
        val vectors = cryptoVectors["key_rotation"] as List<Map<String, Any>>
        for (v in vectors) {
            val baseKey = fromHex(v["base_key"] as String)
            val counter = (v["counter"] as Double).toInt()
            val expected = v["derived_key"] as String

            val derived = KeyRotation.rotateKey(baseKey, counter)
            assertEquals(expected, toHex(derived))
        }
    }
}

File: js/test/cross-validation.test.js
import { describe, it } from 'node:test';
import assert from 'node:assert/strict';
import { readFileSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';

import { sha256, hmacSha256, hkdf, aeadEncrypt, aeadDecrypt, x25519, crc16, crc32, fromHex, toHex } from '../src/index.js';

const __dirname = dirname(fileURLToPath(import.meta.url));
const vectors = JSON.parse(readFileSync(join(__dirname, '../../testdata/crypto_vectors.json'), 'utf-8'));
const hashVectors = JSON.parse(readFileSync(join(__dirname, '../../testdata/hash_vectors.json'), 'utf-8'));

describe('Cross-Validation: SHA-256', () => {
  for (const v of hashVectors.sha256) {
    it(`sha256("${v.input}")`, () => {
      const input = v.hex_input ? fromHex(v.hex_input) : new TextEncoder().encode(v.input);
      const result = toHex(sha256(input));
      assert.equal(result, v.output);
    });
  }
});

describe('Cross-Validation: CRC-16-IBM', () => {
  for (const v of hashVectors.crc16_ibm.vectors) {
    it(`crc16(${v.hex_input || '""'})`, () => {
      const input = v.hex_input ? fromHex(v.hex_input) : new Uint8Array(0);
      const result = crc16(input);
      const expected = parseInt(v.output, 16);
      assert.equal(result, expected);
    });
  }
});

describe('Cross-Validation: CRC-32-IEEE', () => {
  for (const v of hashVectors.crc32_ieee.vectors) {
    it(`crc32(${v.hex_input || '""'})`, () => {
      const input = v.hex_input ? fromHex(v.hex_input) : new Uint8Array(0);
      const result = crc32(input);
      const expected = parseInt(v.output, 16);
      assert.equal(result, expected);
    });
  }
});

describe('Cross-Validation: HMAC-SHA256', () => {
  for (const v of hashVectors.hmac_sha256.vectors) {
    it(v.name, () => {
      const key = fromHex(v.key);
      const data = fromHex(v.data);
      const result = toHex(hmacSha256(key, data));
      assert.equal(result, v.output);
    });
  }
});

describe('Cross-Validation: ChaCha20-Poly1305', () => {
  for (const [i, v] of vectors.chacha20_poly1305.entries()) {
    it(`vector ${i}`, () => {
      const key = fromHex(v.key);
      const nonce = fromHex(v.nonce);
      const plaintext = fromHex(v.plaintext);
      const aad = fromHex(v.aad);
      const expectedCiphertext = fromHex(v.ciphertext);

      // Encrypt and compare
      const encrypted = aeadEncrypt(key, nonce, plaintext, aad);
      assert.equal(toHex(encrypted), v.ciphertext);

      // Decrypt and verify round-trip
      const decrypted = aeadDecrypt(key, nonce, expectedCiphertext, aad);
      assert.equal(toHex(decrypted), v.plaintext);
    });
  }
});

describe('Cross-Validation: HKDF-SHA256', () => {
  for (const [i, v] of vectors.hkdf_sha256.entries()) {
    it(`vector ${i} (length=${v.length})`, () => {
      const ikm = fromHex(v.ikm);
      const salt = v.salt ? fromHex(v.salt) : null;
      const info = v.info ? fromHex(v.info) : null;
      const result = hkdf(salt, ikm, info, v.length);
      assert.equal(toHex(result), v.okm);
    });
  }
});

describe('Cross-Validation: X25519', () => {
  for (const [i, v] of vectors.x25519.entries()) {
    it(`vector ${i}: Alice ECDH(Bob)`, () => {
      const alicePrivate = fromHex(v.alice_private);
      const bobPublic = fromHex(v.bob_public);
      const shared = x25519(alicePrivate, bobPublic);
      assert.equal(toHex(shared), v.shared_secret);
    });

    it(`vector ${i}: Bob ECDH(Alice)`, () => {
      const bobPrivate = fromHex(v.bob_private);
      const alicePublic = fromHex(v.alice_public);
      const shared = x25519(bobPrivate, alicePublic);
      assert.equal(toHex(shared), v.shared_secret);
    });
  }
});

describe('Cross-Validation: Key Rotation', () => {
  for (const [i, v] of vectors.key_rotation.entries()) {
    it(`vector ${i}: counter=${v.counter}`, () => {
      const baseKey = fromHex(v.base_key);
      // Key rotation: HKDF(IKM=baseKey, salt="rotate", info=counter_be32, length=32)
      const info = new Uint8Array(4);
      const dv = new DataView(info.buffer);
      dv.setUint32(0, v.counter, false); // big-endian
      const salt = new TextEncoder().encode('rotate');
      const derived = hkdf(salt, baseKey, info, 32);
      assert.equal(toHex(derived), v.derived_key);
    });
  }
});
